#588 / game configuration

vdedios opened the pull request:
closes #558 
closes #506
closes #492

This PR implements the mechanism (frontend and backend) to run different types of games. It includes classic, misteryZone (black area in the middle of the court) and short paddles. More could be done but this was good for this PR

❗ Also, I've included an upgrade to 60FPS to the backend server and an alert which displays a snackbar if we have a lag bigger than 100ms to warn users about their bad connection. I'm closing for now #492 since the idea of that ticket was to reflect subject's requirements, but we can open an exploratory ticket to investigate how to improve online gaming experience as we talked.

vdedios commented on webapp/src/shared/components/Game/Game.tsx:
```diff
@@ -85,31 +147,62 @@ const Play = ({ gameId }: GameProps) => {
+      },
+      title: 'Classic game',
+      key: 'Classic game',
+      iconVariant: IconVariant.PLAY,
```
I totally made up this icons, this should be changed but I didn't want to lose time looking for pretty ones

vdedios commented on webapp/src/shared/components/Game/hooks/useOnlineGame.tsx:
```diff
@@ -79,7 +98,14 @@ export function useOnlineGame(gameId: string) {
+      const now = Date.now();
+      if (updateTimestamp.current) {
+        now - updateTimestamp.current > MAX_CONNECTION_LAG_IN_MS &&
+          warn('slow connection', 'top');
```
This is good for now, it may be changed in another ticket to a less invasive warning. That being said I think the smallest problem for a player when is experiencing a laggy connection is to see a notification, he/she cannot play indeed.

Damrod replied:
Last time we played, the warning was also showing from the top, and it was very bothersome (more than just the lag) because I couldn't see the opponent's paddle.
But I understand this may be for a different PR.

vdedios replied:
It really depends also on the screen heigh, I'd say with the score component should be less invasive. But, of course, this can be improved later

Damrod commented on transcendence-app/src/game/game.service.ts:
```diff
@@ -92,6 +96,19 @@ export class GameService {
+        return GameModeEnum.shortPaddle;
+      case 'misteryZone':
+        return GameModeEnum.misteryZone;
+      default:
```
What happened with the idea of the non arkanoid rebound?

Damrod commented:
After I configure with one player and am ready, the other player is redirected to a blank page. Maybe something is not loading, or is throwing an exception....

Damrod commented:
I haven't managed to see that issue again.
On the other hand, there's the hidden area problem:
I'd expect the area to be centered. Something like this:
![good](https://user-images.githubusercontent.com/2774420/221152620-a602ed22-3930-4fb8-ae4a-1f97ca95dc79.png)

Instead it's something like this (totally off center):
![bad](https://user-images.githubusercontent.com/2774420/221152673-e26e6768-8c57-42b0-a6de-d5e9eaeb3428.png)

Damrod commented:
I believed we had fixed the problem of the paddle that embeds itself into the walls. What happened?

vdedios commented on transcendence-app/src/game/enums/game-mode.enum.ts:
```diff
@@ -1,4 +1,7 @@
   training = 'training',
   classic = 'classic',
+  shortPaddle = 'shortPaddle',
+  misteryZone = 'misteryZone',
```
ouch! fixed everywhere

vdedios commented on webapp/src/shared/components/Game/hooks/useOnlineGame.tsx:
```diff
@@ -79,7 +98,14 @@ export function useOnlineGame(gameId: string) {
+      if (updateTimestamp.current) {
+        now - updateTimestamp.current > MAX_CONNECTION_LAG_IN_MS &&
+          warn('slow connection', 'top');
+        updateTimestamp.current = now;
+      }
```
you are damn right, fixed!

vdedios commented on webapp/src/shared/components/Game/Game.tsx:
```diff
@@ -85,31 +147,62 @@ const Play = ({ gameId }: GameProps) => {
 };
 
-const Wait = () => {
+const ConfigureGame = ({ submitGameConfig, joinGame }: ConfigureGameProps) => {
```
done

vdedios commented on transcendence-app/src/game/validators/index.ts:
```diff
@@ -30,3 +33,24 @@ export function IsGameCommand() {
+      target: object.constructor,
+      propertyName: propertyName,
+      options: {
+        message: 'The value passed as command is not a valid game command',
```
thanks!

vdedios commented on transcendence-app/src/game/game.service.ts:
```diff
@@ -28,15 +30,17 @@ import { PaginationWithSearchQueryDto } from '../shared/dtos/pagination-with-sea
 type PlayerId = string;
 
-const FPS = 30;
+const FPS = 60;
```
🤞

vdedios commented:
> I believed we had fixed the problem of the paddle that embeds itself into the walls. What happened?

It got lost in the PR I closed, I've added it back. Thanks!

vdedios commented:
> I'd expect the area to be centered. Something like this:

Done, thanks!
